src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -27,14 +31,123 @@
+ * After that, SQL statetemens for re-creating metadata about mx distributed
+ * tables are sent to the worker. Finally, the hasmetadata column of the target node in
+ * pg_dist_node is marked as true.
+ */
+Datum
+sync_metadata_to_node(PG_FUNCTION_ARGS)
+{
+	text *nodeName = PG_GETARG_TEXT_P(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+	char *nodeNameString = text_to_cstring(nodeName);
+	char *extensionOwner = CitusExtensionOwnerName();
+
+	WorkerNode *workerNode = NULL;
+	char *localGroupIdUpdateCommand = NULL;
+	List *recreateMetadataSnapshotCommandList = NIL;
+	List *dropMetadataCommandList = NIL;
+	List *createMetadataCommandList = NIL;
+
+	EnsureSuperUser();
+
+	PreventTransactionChain(true, "sync_metadata_to_node");
+
+	workerNode = FindWorkerNode(nodeNameString, nodePort);
+
+	if (workerNode == NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("you cannot sync metadata to a non-existent node"),
+						errhint("First, add the node with SELECT master_add_node(%s,%d)",
+								nodeNameString, nodePort)));
+	}
+
+	/* generate and add the local group id's update query */
+	localGroupIdUpdateCommand = LocalGroupIdUpdateCommand(workerNode->groupId);
+
+	/* generate the queries which drop the metadata */
+	dropMetadataCommandList = MetadataDropCommands();
+
+	/* generate the queries which create the metadata from scratch */
+	createMetadataCommandList = MetadataCreateCommands();
+
+	recreateMetadataSnapshotCommandList = lappend(recreateMetadataSnapshotCommandList,
+												  localGroupIdUpdateCommand);
+	recreateMetadataSnapshotCommandList = list_concat(recreateMetadataSnapshotCommandList,
+													  dropMetadataCommandList);
+	recreateMetadataSnapshotCommandList = list_concat(recreateMetadataSnapshotCommandList,
+													  createMetadataCommandList);
+
+	/*
+	 * Send the snapshot recreation commands in a single remote transaction and
+	 * error out in any kind of failure. Note that it is not required to send
+	 * createMetadataSnapshotCommandList in the same transaction that we send
+	 * nodeDeleteCommand and nodeInsertCommand commands below.
+	 */
+	SendCommandListToWorkerInSingleTransaction(nodeNameString, nodePort, extensionOwner,
+											   recreateMetadataSnapshotCommandList);
+
+	MarkNodeHasMetadata(nodeNameString, nodePort, true);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * stop_sync_metadata_to_node function sets the hasmetadata column of the specified node
+ * to false in pg_dist_node table, thus indicating that the specified worker node does not
+ * receive DDL changes anymore and cannot be used for issuing mx queries.
+ */
+Datum
+stop_sync_metadata_to_node(PG_FUNCTION_ARGS)
+{
+	text *nodeName = PG_GETARG_TEXT_P(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+	char *nodeNameString = text_to_cstring(nodeName);
+	WorkerNode *workerNode = NULL;
+
+	EnsureSuperUser();
+
+	PreventTransactionChain(true, "stop_sync_metadata_to_node");
```

Comment:
This doesn't seem necessary. MarkNodeHasMetadata can safely be called from a transaction.